### PR TITLE
DOC: Array API : Directory restructure and code cleanup

### PR DESCRIPTION
**Pull request by kritisingh1**

### CODE REVIEW

**Comment by mattip** [comment]
In `doc/source/reference/index.rst`, line 27 `c-api` must now be changed to `c-api/c-api` as the top level link to the moved file. This is causing the build to fail. You might want to rename `c-api.rst` to `index.rst` as well (then the new link becomes `c-api/index` of course).

File: doc/source/reference/c-api/array.rst [hunk 1]
```diff
@@ -2052,6 +2052,17 @@ Calculation
...
 
+
+.. note::
+
+    The out argument specifies where to place the result. If out is
+    NULL, then the output array is created, otherwise the output is
+    placed in out which must be the correct size and type. A new
+    reference to the output array is always returned even when out
+    is not NULL. The caller of the routine has the responsibility
+    to ``DECREF`` out if not NULL or a memory-leak will occur.
```
Comment (eric-wieser) [inline doc/source/reference/c-api/array.rst]: nit: `` `Py_DECREF` ``

File: doc/source/reference/c-api/array.rst [hunk 2]
```diff
@@ -2655,22 +2654,22 @@ cost of a slight overhead.
...
-    * NPY_NEIGHBORHOOD_ITER_CONSTANT_PADDING: constant padding. Outside bounds
+    * **NPY_NEIGHBORHOOD_ITER_CONSTANT_PADDING**: constant padding. Outside bounds
       values will be the same as the first item in fill_value.
-    * NPY_NEIGHBORHOOD_ITER_MIRROR_PADDING: mirror padding. Outside bounds
+    * **NPY_NEIGHBORHOOD_ITER_MIRROR_PADDING**: mirror padding. Outside bounds
       values will be as if the array items were mirrored. For example, for the
       array [1, 2, 3, 4], x[-2] will be 2, x[-2] will be 1, x[4] will be 4,
       x[5] will be 1, etc...
-    * NPY_NEIGHBORHOOD_ITER_CIRCULAR_PADDING: circular padding. Outside bounds
+    * **NPY_NEIGHBORHOOD_ITER_CIRCULAR_PADDING**: circular padding. Outside bounds
```
Comment (eric-wieser) [inline doc/source/reference/c-api/array.rst]: These would be better as
```
.. c:macro:: NPY_NEIGHBORHOOD_ITER_CIRCULAR_PADDING
    circular padding. Outside bounds ...
```

File: doc/source/reference/c-api/array.rst [hunk 3]
```diff
@@ -1578,7 +1578,7 @@ Flag checking
 ^^^^^^^^^^^^^
 
 For all of these macros *arr* must be an instance of a (subclass of)
-:c:data:`PyArray_Type`, but no checking is done.
```
Comment (eric-wieser) [inline doc/source/reference/c-api/array.rst]: Is this change indicating that checking _is_ now done?

File: doc/source/reference/c-api/array.rst [hunk 4]
```diff
@@ -2655,22 +2654,22 @@ cost of a slight overhead.
...
       array [1, 2, 3, 4], x[-2] will be 2, x[-2] will be 1, x[4] will be 4,
       x[5] will be 1, etc...
-    * NPY_NEIGHBORHOOD_ITER_CIRCULAR_PADDING: circular padding. Outside bounds
+    * **NPY_NEIGHBORHOOD_ITER_CIRCULAR_PADDING**: circular padding. Outside bounds
       values will be as if the array was repeated. For example, for the
       array [1, 2, 3, 4], x[-2] will be 3, x[-2] will be 4, x[4] will be 1,
       x[5] will be 2, etc...
 
-    If the mode is constant filling (NPY_NEIGHBORHOOD_ITER_CONSTANT_PADDING),
+    If the mode is constant filling (**NPY_NEIGHBORHOOD_ITER_CONSTANT_PADDING**),
```
Comment (eric-wieser) [inline doc/source/reference/c-api/array.rst]: `` `NPY_NEIGHBORHOOD_ITER_CONSTANT_PADDING` `` will produce a link

File: doc/source/reference/c-api/array.rst (hunk 2 above)
Comment (seberg) [reply doc/source/reference/c-api/array.rst]: Yeah, sorry, my bad probably.

File: doc/source/reference/c-api/array.rst (hunk 3 above)
Comment (seberg) [reply doc/source/reference/c-api/array.rst]: Dunno, I think we read must as the user having to take care of it, so it seemed a bit duplicate. But maybe it is not bad to repeat it.

**Comment by seberg** [comment]
Thanks, lets put this in (there is the small question if the "must" is strong enough to say that it is not actually checked, but I feel it is a minor discussion).

**Pull request merged**